src/auth/stscredentials.ts

```diff
@@ -0,0 +1,179 @@
+ */
+export class StsCredentials extends OAuthClientAuthHandler {
+  private transporter: DefaultTransporter;
+
+  /**
+   * Initializes an STS credentials instance.
+   * @param tokenExchangeEndpoint The token exchange endpoint.
+   * @param clientAuthentication The client authentication credentials if
+   *   available.
+   */
+  constructor(
+    private readonly tokenExchangeEndpoint: string,
+    clientAuthentication?: ClientAuthentication
+  ) {
+    super(clientAuthentication);
+    this.transporter = new DefaultTransporter();
+  }
+
+  /**
+   * Exchanges the provided token for another type of token based on the
+   * rfc8693 spec.
+   * @param stsCredentialsOptions The token exchange options used to populate
+   *   the token exchange request.
+   * @param additionalHeaders Optional additional headers to pass along the
+   *   request.
+   * @param options Optional additional GCP-specific non-spec defined options
+   *   to send with the request.
+   *   Example: `&options=${encodeUriComponent(JSON.stringified(options))}`
+   * @return A promise that resolves with the token exchange response containing
+   *   the requested token and its expiration time.
+   */
+  async exchangeToken(
+    stsCredentialsOptions: StsCredentialsOptions,
+    additionalHeaders?: Headers,
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    options?: {[key: string]: any}
+  ): Promise<StsSuccessfulResponse> {
+    const values: StsRequestOptions = {
+      grant_type: stsCredentialsOptions.grantType,
+      resource: stsCredentialsOptions.resource,
+      audience: stsCredentialsOptions.audience,
+      scope: stsCredentialsOptions.scope?.join(' '),
+      requested_token_type: stsCredentialsOptions.requestedTokenType,
+      subject_token: stsCredentialsOptions.subjectToken,
+      subject_token_type: stsCredentialsOptions.subjectTokenType,
+      actor_token: stsCredentialsOptions.actingParty?.actorToken,
+      actor_token_type: stsCredentialsOptions.actingParty?.actorTokenType,
+      // Non-standard GCP-specific options.
+      options: options && JSON.stringify(options),
+    };
+    // Remove undefined fields.
+    Object.keys(values).forEach(key => {
+      if (typeof values[key] === 'undefined') {
+        delete values[key];
+      }
+    });
+
+    const headers = {
+      'Content-Type': 'application/x-www-form-urlencoded',
+    };
+    // Inject additional STS headers if available.
+    Object.assign(headers, additionalHeaders || {});
+
+    const opts: GaxiosOptions = {
+      url: this.tokenExchangeEndpoint,
+      method: 'POST',
+      headers,
+      data: querystring.stringify(values),
+      responseType: 'json',
+    };
+    // Apply OAuth client authentication.
+    this.applyClientAuthenticationOptions(opts);
+
+    try {
+      const response = await this.transporter.request<StsSuccessfulResponse>(
+        opts
+      );
+      // Successful response.
+      const stsSuccessfulResponse = response.data;
```

Comment:
I'm curious - why return the `response` object along with the data?

Reply:
Mostly following the existing pattern established [here](https://github.com/bojeil-google/google-auth-library-nodejs/blob/a7e5701a8394d79fe93d28794467747a23cf9ff4/src/auth/oauth2client.ts#L309) and elsewhere in that file.